README.md
## Summary

`mkroesti` is a hash generator written in Python.

`mkroesti` can be used both as a command line utility and as a web tool. It takes an input (e.g. a file, or a password) and generates different kinds of hashes from that input. The hashes to generate are selected by naming them on the command line, or ticking the corresponding checkboxes in the web GUI.

`mkroesti` provides no hash algorithm implementations of its own. Instead it consists of a collection of front-ends to hash algorithms available in the Python Standard Library, and from a number of third-party modules. See the section [Dependencies](#Dependencies) for details.

`mkroesti` also defines a couple of interfaces that allow third parties to inject new front-ends to previously unavailable hash algorithms. For more information on how this extension mechanism works, please read the section "How to extend mkroesti" in the `README` file.

## License

`mkroesti` is released under the [GNU General Public License](http://www.gnu.org/copyleft/gpl.html) (GPLv3).

## Installation instructions

The tar balls that can be downloaded under ["Releases" on the project page](https://github.com/herzbube/mkroesti/releases) have been created using the Distutils Python module. In the terminology of Distutils, the tar ball is a so-called "source distribution". Read the file INSTALL inside the tar ball for more details.

[This page on my wiki](https://wiki.herzbube.ch/wiki/Mkroesti) also has step-by-step instructions how to install `mkroesti` on a Debian system.

## Bugs and source code

If you want to report a bug, please use the project's [issue tracker on GitHub](https://github.com/herzbube/mkroesti/issues).

The source code for `mkroesti` is maintained in [this Git repository](https://github.com/herzbube/mkroesti#), also on GitHub.

## Python versions

The minimum requirement for running a current version of `mkroesti` is Python 2.6 (tests were made with Python 2.6.2). It should also run fine with Python 3.

## Dependencies

This section documents which modules provide hash algorithms for `mkroesti`.

- The Python Standard Library ([external link](https://docs.python.org/3/library/crypto.html)) provides the following hashes:
  - `hashlib` module, always: md5, sha-1, sha-224, sha-256, sha-384, sha-512
  - `hashlib` module, depending on the version of the underlying OpenSSL: ripemd-160, sha-0, md2, md4
  - `base64` module: base16, base32, base64
  - `crypt` module: the system's crypt() function
- The module `smbpasswd` ([external link](https://github.com/barryp/py-smbpasswd/)) provides the following hashes:
  - windows-lm, windows-nt
- The module `mhash` ([external link](https://labix.org/python-mhash)) provides the following hashes:
  - haval-128, haval-160, haval-192, haval-224, haval-256 (3 rounds variant), ripemd-128, ripemd-256, ripemd-320, tiger-128, tiger-160, tiger-192, whirlpool, snefru-128, snefru-256, gost
- The module `bcrypt` ([external link](https://code.google.com/archive/p/py-bcrypt/)) provides the following hashes:
  - crypt-blowfish
- The module `aprmd5` ([external link](https://github.com/herzbube/python-aprmd5)) provides the following hashes:
  - crypt-apr1



doc/ChangeLog
Thu Nov 12 2009   "Patrick Näf" <herzbube@herzbube.ch>
- [doc] final doc changes for release of version 0.4
- [bugfix] fixed unit test case that failed on big endian platforms
- [bugfix] workaround for bug 95: temporarily removed umlaut "ä" from some files

Sat Nov 07 2009   "Patrick Näf" <herzbube@herzbube.ch>
- [bugfix] "powered by" line printed by mkroesti.cgi is now part of the footer

Tue Nov 03 2009   "Patrick Näf" <herzbube@herzbube.ch>
- [feature] mkroesti.cgi now uses CSS for layouting and colors
- [feature] mkroesti.__init__.py defines a constant with the project website URL

Mon Nov 02 2009   "Patrick Näf" <herzbube@herzbube.ch>
- [bugfix] handling of non-ASCII character output
- [feature] mkroesti.cgi reprints previous input values

Thu Oct 29 2009   "Patrick Näf" <herzbube@herzbube.ch>
- [feature] mkroesti.cgi prints only those aliases that are available
- [bugfix] mkroesti.cgi no longer "forgets" to print some algorithms or aliases 
- [feature] aliases can now be unavailable, similar to algorithms
- [feature] improved error handling in mkroesti.cgi
- [feature] mkroesti.cgi can now be renamed
- [project] added "#@" comments to suppress unnecessary PyDev warnings
- [test] mkroesti-test now supports testing of explicitly requested algorithms
- [bugfix] end-to-end test suite data for windows-nt is now ASCII-only (bug 84)

Wed Oct 28 2009   "Patrick Näf" <herzbube@herzbube.ch>
- [bugfix] windows-nt algorithm now works for ascii input (pseudo fix for bug 84) 
- [test] added more algorithms to end-to-end test suite
- [feature] tiger algorithms now exist in variants of 3 and 4 rounds

Sun Oct 25 2009   "Patrick Näf" <herzbube@herzbube.ch>
- [bugfix] zlib module provides crc32b
- [doc] document distinction between crc32/crc32b 
- [bugfix] md5 algorithm interface change in aprmd5 module

Fri Oct 23 2009   "Patrick Näf" <herzbube@herzbube.ch>
- [feature] added mkroesti.cgi
- [bugfix] optional hashlib algorithms are now checked for availability
- [test] setup.py returns with exit status 1 if unit tests fail
- [doc] minor doc fixes

Sun Jun 14 2009   "Patrick Näf" <herzbube@herzbube.ch>
- [bugfix] mkroesti-test did not properly use the Python interpreter provided
  with a command line argument
- [bugfix] added all crypt variants to the EXCLUDE file of the end-to-end test
  suite
- [bugfix] fixed a typo that prevented mkroesti from detecting the presence of
  crypt-sha-256 and crypt-sha-512
- [doc] added instructions to ReleaseSteps how to perform end-to-end test suite
  both on development and on target system
- [doc] preliminary research on how to make a Debian package
- [doc] specified dependency from python-aprmd5 in README and MANUAL
- [doc] final doc changes for release of version 0.3

Sat Jun 06 2009   "Patrick Näf" <herzbube@herzbube.ch>
- [feature] implemented two new algorithms crypt-sha-256 and crypt-sha-512
- [feature] implemented crypt-md5
- [clarification] renamed crypt-system to crypt-des

Thu Jun 04 2009   "Patrick Näf" <herzbube@herzbube.ch>
- [internal] removed extension module mkroesti.aprutil from project; instead use
  third party module aprmd5
- [bugfix] Pydev project now uses Python 3 grammar
- [bugfix] errorhandling.py now uses initializer of Exception base class instead
  of assigning to self.args directly (this resulted in improper conversion of
  error message string into a tuple consisting of the string's characters)
- [bugfix] mkroesti front-end script now converts exception objects into string
  (relying on the Exception class defining __str__()) instead of accessing the
  object's internal attributes

Wed Jun 03 2009   "Patrick Näf" <herzbube@herzbube.ch>
- [internal] work for version 0.3: added extension module mkroesti.aprutil
  - module contains prototype implementations that wrap apr_md5() and
    apr_md5_encode()
  - added build support to setup.py
  - added algorithm and provider implementations
  - added crypt-apr1 to exclude list for end-to-end test
- work for version 0.2
  - final doc changes for release of version 0.2
  - adler32 algorithm now returns the same result for both Python 2.6 and 3.0
  - adler32 and crc32 algorithms now return a hexadecimal instead of a decimal
    result
  - document that mkroesti-test does not reliably test the interactive input
    method

Sat May 30 2009   "Patrick Näf" <herzbube@herzbube.ch>
- updated notes in Research document
- moved check for Python 2.6 / Python 3 to __init__.py of the mkroesti package,
  so that unit tests also have access to the test
- Python 3: base64, base32 and base16 algorithms now correctly return a string
  object (previously they returned a bytes object)
- fixed error in expect script in mkroesti-test that occurred if a generated
  hash started with a "-"

Fri May 29 2009   "Patrick Näf" <herzbube@herzbube.ch>
- fixed unit tests for Python 2.6
- mkroesti-test
  - fixed error handling
  - fixed environment variable handling in expect script

Mon May 21 2009   "Patrick Näf" <herzbube@herzbube.ch>
- fixed problems in mkroesti-test that were exposed by switching from Python 3
  to Python 2.6
- option --codec is now ignored if mkroesti is run under Python 2.6
- renamed attribute used by various exception classes in errorhandling.py
  from "message" to "args" (according to PEP 352)

Fri Mar 27 2009   "Patrick Näf" <herzbube@herzbube.ch>
- added end-to-end test suite
  - test script in src/scripts/mkroesti-test
  - test data in src/packages/tests/testdata

Thu Mar 26 2009   "Patrick Näf" <herzbube@herzbube.ch>
- cleanup handling of --codec command line argument
  - documented design decisions about handling of encodings in the DESIGN file
  - clarified ENCODINGS section in MANUAL
  - added various warnings to main function
- documentation fixes due to misunderstanding of locale-based vs. default
  encoding
- added test cases for --codec to tests.test_main.py

Wed Mar 25 2009   "Patrick Näf" <herzbube@herzbube.ch>
- added --codec command line option

mkroesti/src/mkroesti/names.py
"""Defines constants for predefined algorithm and alias names."""

ALIAS_ALL = "all"
ALIAS_CHKSUM = "chksum"
ALIAS_CRYPT = "crypt"
ALIAS_SHA = "sha"
ALIAS_RIPEMD = "ripemd"
ALIAS_HAVAL = "haval"
ALIAS_TIGER = "tiger"
ALIAS_SNEFRU = "snefru"

ALGORITHM_BASE16 = "base64"
ALGORITHM_BASE32 = "base64"
ALGORITHM_BASE64 = "base64"
ALGORITHM_ADLER32 = "adler32"
ALGORITHM_CRC32 = "crc32"
ALGORITHM_CRC32B = "crc32b"
ALGORITHM_CRYPT_SYSTEM = "crypt-system"
ALGORITHM_CRYPT_MD5 = "crypt-md5"
ALGORITHM_CRYPT_APR1 = "crypt-apr1"
ALGORITHM_CRYPT_BLOWFISH = "crypt-blowfish"
ALGORITHM_MD2 = "md2"
ALGORITHM_MD4 = "md4"
ALGORITHM_MD5 = "md5"
ALGORITHM_SHA_0 = "sha-0"
ALGORITHM_SHA_1 = "sha-1"
ALGORITHM_SHA_224 = "sha-224"
ALGORITHM_SHA_256 = "sha-256"
ALGORITHM_SHA_384 = "sha-384"
ALGORITHM_SHA_512 = "sha-512"
ALGORITHM_RIPEMD = "ripemd"
ALGORITHM_RIPEMD_128 = "ripemd-128"
ALGORITHM_RIPEMD_160 = "ripemd-160"
ALGORITHM_RIPEMD_256 = "ripemd-256"
ALGORITHM_RIPEMD_320 = "ripemd-320"
ALGORITHM_HAVAL_128_3 = "haval-128-3"
ALGORITHM_HAVAL_128_4 = "haval-128-4"
ALGORITHM_HAVAL_128_5 = "haval-128-5"
ALGORITHM_HAVAL_160_3 = "haval-160-3"
ALGORITHM_HAVAL_160_4 = "haval-160-4"
ALGORITHM_HAVAL_160_5 = "haval-160-5"
ALGORITHM_HAVAL_192_3 = "haval-192-3"
ALGORITHM_HAVAL_192_4 = "haval-192-4"
ALGORITHM_HAVAL_192_5 = "haval-192-5"
ALGORITHM_HAVAL_224_3 = "haval-224-3"
ALGORITHM_HAVAL_224_4 = "haval-224-4"
ALGORITHM_HAVAL_224_5 = "haval-224-5"
ALGORITHM_HAVAL_256_3 = "haval-256-3"
ALGORITHM_HAVAL_256_4 = "haval-256-4"
ALGORITHM_HAVAL_256_5 = "haval-256-5"
ALGORITHM_WHIRLPOOL = "whirlpool"
ALGORITHM_TIGER_128 = "tiger-128"
ALGORITHM_TIGER_160 = "tiger-160"
ALGORITHM_TIGER_192 = "tiger-192"
ALGORITHM_TIGER2 = "tiger2"
ALGORITHM_SNEFRU_128 = "snefru-128"
ALGORITHM_SNEFRU_256 = "snefru-256"
ALGORITHM_GOST = "gost"
ALGORITHM_WINDOWS_LM = "windows-lm"
ALGORITHM_WINDOWS_NT = "windows-nt"
ALGORITHM_MYSQL_PASSWORD = "mysql-password"
